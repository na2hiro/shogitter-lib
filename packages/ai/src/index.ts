export * from "./playerRegistry.js";
export { loadModuleOnServer } from "./players/all.js";
export type * from "./Player.js";
