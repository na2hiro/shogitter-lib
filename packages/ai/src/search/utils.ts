import { ShogitterAiException } from "../ShogitterAiException.js";

export function randomSelect<T>(arr: T[]) {
  if (arr.length === 0)
    throw new ShogitterAiException("randomSelect: empty array");
  return arr[Math.floor(Math.random() * arr.length)];
}
