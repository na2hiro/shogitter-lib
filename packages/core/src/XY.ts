import { ShogitterCoreException } from "./utils/phpCompat.js";
import { num2kan_decimal } from "./MyLib.js";
import { Direction } from "./Direction.js";

export enum RelativeType {
  TATEYOKO = 1,
  ALLROUND = 2,
  HALFROUND = 3,
}

export const XYMatcher = (x: number, y: number) => (xy: XY) =>
  xy.x == x && xy.y == y;

export type XYObj = [number, number];

export class RelXY {
  x: number;
  y: number;
  static constants = {
    [RelativeType.TATEYOKO]: [
      //縦横１マスの位置（計４マス）
      new RelXY(1, 0),
      new RelXY(-1, 0),
      new RelXY(0, 1),
      new RelXY(0, -1),
    ],
    [RelativeType.ALLROUND]: [
      //周囲１マスの位置（計８マス）
      new RelXY(1, 0),
      new RelXY(-1, 0),
      new RelXY(0, 1),
      new RelXY(0, -1),
      new RelXY(1, 1),
      new RelXY(1, -1),
      new RelXY(-1, 1),
      new RelXY(-1, -1),
    ],
    [RelativeType.HALFROUND]: [
      //周囲１マスの位置を半周（計４マス）
      new RelXY(1, 0),
      new RelXY(1, 1),
      new RelXY(0, 1),
      new RelXY(-1, 1),
    ],
  };
  static getVector(type: RelativeType): RelXY[] {
    return RelXY.constants[type];
  }

  static fromArray([x, y]: XYObj) {
    return new RelXY(x, y);
  }

  constructor(x: number, y: number) {
    this.x = x;
    this.y = y;
  }

  toString() {
    return `(${this.x}, ${this.y}) `;
  }

  getHash() {
    return `${this.x},${this.y}`;
  }

  add(x: number, y: number) {
    this.x += x;
    this.y += y;
  }

  addRel(relXY: RelXY, tani = 1) {
    this.x += relXY.x * tani;
    this.y += relXY.y * tani;
  }

  set(x: number, y: number) {
    this.x = x;
    this.y = y;
  }

  equals(xy: XY) {
    return this.x == xy.x && this.y == xy.y;
  }

  getClone(x = 0, y = 0) {
    const tmp = new RelXY(this.x, this.y);
    tmp.add(x, y);
    return tmp;
  }

  getCloneRel(relXY: RelXY, tani = 1) {
    const tmp = new RelXY(this.x, this.y);
    tmp.addRel(relXY, tani);
    return tmp;
  }

  getArray(): XYObj {
    return [this.x, this.y];
  }

  turn(direction: Direction) {
    let x, y;
    switch (direction) {
      case Direction.BLACK: //先手0度回転
        x = this.x;
        y = this.y;
        break;
      case Direction.WHITE: //後手180度回転
        x = -this.x;
        y = -this.y;
        break;
      case Direction.LEFT: //左手90度回転
        x = this.y;
        y = -this.x;
        break;
      case Direction.RIGHT: //右手270度回転
        x = -this.y;
        y = this.x;
        break;
      default:
        throw new ShogitterCoreException(
          `駒の向き${direction}が定義されていません。`
        );
    }
    this.x = x;
    this.y = y;
  }

  turnReverse(direction: Direction) {
    let x, y;
    switch (direction) {
      case Direction.BLACK: //先手0度回転
        x = this.x;
        y = this.y;
        break;
      case Direction.WHITE: //後手180度回転
        x = -this.x;
        y = -this.y;
        break;
      case Direction.LEFT: //左手270度回転
        x = -this.y;
        y = this.x;
        break;
      case Direction.RIGHT: //右手90度回転
        x = this.y;
        y = -this.x;
        break;
      default:
        throw new ShogitterCoreException(
          `駒の向き${direction}が定義されていません。`,
          1
        );
    }
    this.x = x;
    this.y = y;
  }

  //単位ベクトル化(最小公倍数で割る)
  unify() {
    let abs;
    if (this.x == 0) {
      if (this.y == 0) {
        return 1;
      } else {
        abs = Math.abs(this.y);
        this.y /= abs;
        return abs;
      }
    } else {
      if (this.y == 0) {
        abs = Math.abs(this.x);
        this.x /= abs;
        return abs;
      } else {
        const gcd = RelXY.gcd(Math.abs(this.x), Math.abs(this.y));
        this.x /= gcd;
        this.y /= gcd;
        return gcd;
      }
    }
  }

  //最大公倍数を求める（ユークリッドの互除法）
  static gcd(m: number, n: number) {
    while (m % n != 0) {
      const temp = n;
      n = m % n;
      m = temp;
    }
    return n;
  }

  getCloneFlipped() {
    const ret = this.getClone();
    ret.x *= -1;
    ret.y *= -1;
    return ret;
  }
}

export default class XY extends RelXY {
  /**
   * 座標を生成
   * @param x
   * @param y
   */
  constructor(x: number, y: number) {
    super(x, y);
  }

  static fromArray([x, y]: XYObj) {
    return new XY(x, y);
  }

  toArray(): XYObj {
    return [this.x, this.y];
  }

  toString() {
    return `(${this.x}, ${this.y})`;
  }

  getFormat() {
    return this.x + num2kan_decimal(`${this.y}`);
  }

  turn() {
    throw new ShogitterCoreException("座標が回転できません。");
  }

  turnReverse() {
    throw new ShogitterCoreException("座標が回転できません。");
  }

  getDistance(xy: XY) {
    return Math.pow(this.x - xy.x, 2) + Math.pow(this.y - xy.y, 2);
  }

  to(xy: XY) {
    return new RelXY(xy.x - this.x, xy.y - this.y);
  }

  from(xy: XY) {
    return new RelXY(this.x - xy.x, this.y - xy.y);
  }

  clone() {
    return new XY(this.x, this.y);
  }

  getClone(x = 0, y = 0): XY {
    const tmp = new XY(this.x, this.y);
    tmp.add(x, y);
    return tmp;
  }

  getCloneRel(relXY: RelXY, tani = 1) {
    const tmp = new XY(this.x, this.y);
    tmp.addRel(relXY, tani);
    return tmp;
  }
}
