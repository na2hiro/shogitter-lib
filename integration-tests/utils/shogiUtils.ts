import Shogi from "../../src/Shogi";
import XY from "../../src/XY";
import {Species} from "../../src/Ban";
import {serialize} from "./ShogiSerializer";
import {Direction} from "../../src/Direction";

export function move(shogi: Shogi, fromX: number, fromY: number, toX: number, toY: number, nari = false) {
    try {
        rawMove(shogi, fromX, fromY, toX, toY, nari);
    } catch (e) {
        console.error(serialize(shogi, null, null, null, null, null))
        throw e;
    }
}

export function rawMove(shogi: Shogi, fromX: number, fromY: number, toX: number, toY: number, nari = false) {
    shogi.runCommand({
        type: "move",
        from: [fromX, fromY],
        to: [toX, toY],
        nari
    })
}

export function put(shogi: Shogi, toX: number, toY: number, species: Species, id?: number) {
    try {
        rawPut(shogi, toX, toY, species, id);
    } catch (e) {
        console.error(serialize(shogi, null, null, null, null, null))
        throw e;
    }
}

export function rawPut(shogi: Shogi, toX: number, toY: number, species: Species, id?: number) {
    shogi.runCommand({
        type: "put",
        to: [toX, toY],
        put: species,
        direction: shogi.teban.getNowDirection(),
        id
    })
}

export function assertCellIsNull(shogi: Shogi, x: number, y: number) {
    expect(shogi.ban.get(new XY(x, y)).isNull()).toBeTruthy;
}

export function assertCell(shogi: Shogi, x: number, y: number, direction: Direction, species: Species) {
    const koma = shogi.ban.get(new XY(x, y));
    expect(koma.isNull()).toBeFalsy();
    expect(koma.species).toBe(species);
    expect(koma.direction).toBe(direction);
}
