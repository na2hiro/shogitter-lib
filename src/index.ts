import Shogi from "./Shogi";
import {Direction} from "./Direction";
import {Species} from "./Ban";
import { PlayerInfo } from "./Teban";
import { KomaInfo, shogitterDB } from "./ShogitterDB";
import XY from "./XY";

export * from "./Shogi";
export default Shogi;
export {Direction, Species, KomaInfo, PlayerInfo, shogitterDB, XY};