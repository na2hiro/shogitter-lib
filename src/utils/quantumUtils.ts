import { RelXY } from "../XY";
import { QuantumData } from "../strategy/MoveEffectStrategy";
import { Direction } from "../Direction";
import { ShogitterCoreException } from "./phpCompat";

export type QuantumPiece =
  | "Fu"
  | "Ky"
  | "Ke"
  | "Gi"
  | "Ki"
  | "Ka"
  | "Hi"
  | "Ou";
const allPieces: QuantumPiece[] = [
  "Fu",
  "Ky",
  "Ke",
  "Gi",
  "Ki",
  "Ka",
  "Hi",
  "Ou",
];
const initialPieces: { [piece in QuantumPiece]: number } = {
  Fu: 9,
  Ky: 2,
  Ke: 2,
  Gi: 2,
  Ki: 2,
  Ka: 1,
  Hi: 1,
  Ou: 1,
};
type Move = [number, number];
const kin: Move[] = [
  [1, -1],
  [0, -1],
  [-1, -1],
  [1, 0],
  [-1, 0],
  [0, 1],
];
const moves: { [piece in QuantumPiece]: Move[] }[] = [
  {
    Fu: [[0, -1]],
    Ky: [
      [0, -1],
      [0, -2],
    ],
    Ke: [
      [-1, -2],
      [1, -2],
    ],
    Gi: [
      [1, -1],
      [0, -1],
      [-1, -1],
      [1, 1],
      [-1, 1],
    ],
    Ki: kin,
    Ka: [
      [1, -1],
      [-1, -1],
      [1, 1],
      [-1, 1],
      [2, -2],
      [-2, -2],
      [2, 2],
      [-2, 2],
    ],
    Hi: [
      [1, 0],
      [-1, 0],
      [0, 1],
      [0, -1],
      [2, 0],
      [-2, 0],
      [0, 2],
      [0, -2],
    ],
    Ou: [
      [1, -1],
      [-1, -1],
      [1, 1],
      [-1, 1],
      [1, 0],
      [-1, 0],
      [0, 1],
      [0, -1],
    ],
  },
  {
    Fu: kin,
    Ky: kin,
    Ke: kin,
    Gi: kin,
    Ki: [],
    Ka: [
      [1, 0],
      [-1, 0],
      [0, 1],
      [0, -1],
      [1, -1],
      [-1, -1],
      [1, 1],
      [-1, 1],
      [2, -2],
      [-2, -2],
      [2, 2],
      [-2, 2],
    ],
    Hi: [
      [1, -1],
      [-1, -1],
      [1, 1],
      [-1, 1],
      [1, 0],
      [-1, 0],
      [0, 1],
      [0, -1],
      [2, 0],
      [-2, 0],
      [0, 2],
      [0, -2],
    ],
    Ou: [],
  },
];
const movingBoards: { [vec: string]: SuperPiece }[] = [{}, {}];
[0, 1].forEach((promoted) => {
  allPieces.map((piece) => {
    moves[promoted][piece].forEach((move) => {
      const vec = move[0] + "," + move[1];
      if (!movingBoards[promoted][vec]) movingBoards[promoted][vec] = [];
      movingBoards[promoted][vec].push(piece);
    });
  });
});

export enum MoveType {
  Promote,
  NoPromote1,
  NoPromote2,
  Captured,
  Normal,
}

export type Result = {
  kinds: SuperPiece[];
  fulls: QuantumPiece[];
};
type SuperPiece = QuantumPiece[];
type Possibilities = Possibility[];
type Possibility = {
  piece: SuperPiece;
  indices: number[];
};

export type QuantumMove = {
  vec: RelXY;
  promote: boolean;
};
export const runQuantum = (
  position: number,
  move: QuantumMove | null,
  moveType: MoveType,
  kinds: SuperPiece[]
) => {
  const superr = move == null ? allPieces.slice() : move2superPiece(move);
  const sp = filterByType(moveType, superr);
  const newKinds = kinds.slice();
  if (position < kinds.length) {
    newKinds[position] = kinds[position].filter(
      (kind) => sp.indexOf(kind) >= 0
    );
  } else {
    newKinds.push(sp);
  }
  return getResult(newKinds);
};

export const findQuantumMochigoma = (
  data: QuantumData[],
  direction: Direction,
  id: number
) => {
  for (let i = 0; i < 2; i++) {
    for (let j = 0; j < data[i].xys.length; j++) {
      if (data[i].xys[j] == "mochi" + direction + "-" + id)
        return data[i].d.kinds[j];
    }
  }

  throw new ShogitterCoreException(
    `quantum mochigoma not found for dir=${direction}, id=${id}`
  );
};
export const findQuantumKoma = (
  data: QuantumData[] | undefined,
  x: number,
  y: number,
  direction: Direction
) => {
  if (!data) return allPieces.slice();

  for (let i = 0; i < 2; i++) {
    for (let j = 0; j < data[i].xys.length; j++) {
      if (data[i].xys[j] == `(${x}, ${y})`) {
        return data[i].d.kinds[j];
      }
    }
  }
  // Haven't moved yet
  return allPieces.filter((s) => data[direction!].d.fulls.indexOf(s) == -1);
};

export const stringifyQuantumPiece = (
  piece: QuantumPiece,
  promoted = false
) => {
  if (promoted) {
    return {
      Fu: "と",
      Ky: "杏",
      Ke: "圭",
      Gi: "全",
      Ki: "",
      Ka: "馬",
      Hi: "龍",
      Ou: "",
    }[piece];
  }
  return {
    Fu: "歩",
    Ky: "香",
    Ke: "桂",
    Gi: "銀",
    Ki: "金",
    Ka: "角",
    Hi: "飛",
    Ou: "玉",
  }[piece];
};

function checkFromSuperPiece(pieces: SuperPiece[]) {
  // superPiece2union
  const superPiece2Indices: { [superPiece: string]: number[] } = {};
  pieces.map((piece, index) => {
    if (piece.length == 0)
      throw new Error(`Invalid move combination at ${index}`);
    const p = piece.slice().sort().join(",");
    if (!superPiece2Indices[p]) superPiece2Indices[p] = [];
    superPiece2Indices[p].push(index);
  });
  let possibilities: Possibilities = [];
  for (let superPiece in superPiece2Indices) {
    possibilities.push({
      piece: superPiece.split(",") as QuantumPiece[],
      indices: superPiece2Indices[superPiece],
    });
  }
  const allNonemptySubsets = getAllSubsets(possibilities).slice(1);
  // countsUnionWithIndices
  return allNonemptySubsets
    .map((subset) => {
      const pieceSet = new Set<QuantumPiece>();
      let indicesList: number[] = [];
      subset.forEach(({ piece, indices }) => {
        piece.forEach((p) => pieceSet.add(p));
        indicesList = indicesList.concat(indices);
      });
      return { piece: Array.from(pieceSet), indices: indicesList };
    })
    .map(({ piece, indices }) => {
      const max = calcMax(piece);
      // console.log(max, piece, piece.length, indices, indices.length)
      if (indices.length < max) return { piece: [], indices };
      if (indices.length == max) return { piece, indices };
      throw new Error(`Piece exhausted ${piece} (max ${max}) & ${indices}`);
    });
}

function calcMax(superPiece: SuperPiece) {
  let max = 0;
  for (let [kind, amount] of Object.entries(initialPieces)) {
    if (superPiece.indexOf(kind as QuantumPiece) < 0) continue;
    max += amount;
  }
  return max;
}

const getAllSubsets: <T>(arr: T[]) => T[][] = (theArray) =>
  theArray.reduce(
    (subsets, value) => subsets.concat(subsets.map((set) => [value, ...set])),
    [[]]
  );
type Assign = {
  fst: Possibilities;
  snd: Possibilities;
};

function assign2list(length: number, { fst, snd }: Assign) {
  const unwinded = fst
    .concat(snd)
    .reduce(
      (acc, { piece, indices }) =>
        acc.concat(indices.map((i) => ({ piece, i }))),
      [] as { piece: SuperPiece; i: number }[]
    );
  // TODO: nub
  unwinded.sort((a, b) => a.i - b.i);
  let ret: string[][] = [];
  unwinded.forEach(({ piece, i }) => {
    ret[i] = piece;
  });
  for (let i = 0; i < length; i++) {
    if (!ret[i]) ret[i] = [];
  }
  return ret;
}

function assign(details: Possibilities): Assign {
  const details2 = details.filter((detail) => detail.piece.length > 0);
  const res = assign1(details2);
  if (!res) {
    return { fst: [], snd: details2 };
  }
  const { fst, snd } = assign(res.rest);
  return { fst: fst.concat(res.found), snd: snd };
}

type Assign1 = {
  found: Possibility;
  rest: Possibilities;
};

function assign1(xs: Possibilities): Assign1 {
  const found = xs.find(({ piece: x }) => {
    return xs.every(
      ({ piece: xx }) => pieceEquals(x, xx) || !isSubsetOf(xx, x)
    );
  });
  if (!found) return null;
  return {
    found,
    rest: xs
      .filter((x) => !pieceEquals(x.piece, found.piece))
      .map(({ piece, indices }) => ({
        piece: piece.filter((kind) => found.piece.indexOf(kind) < 0),
        indices: indices.filter((i) => found.indices.indexOf(i) < 0),
      })),
  };
}

function isSubsetOf<T>(xs: T[], ys: T[]) {
  return xs.every((x) => ys.indexOf(x) >= 0);
}

function pieceEquals(piece1: SuperPiece, piece2: SuperPiece) {
  return piece1.slice().sort().join(",") === piece2.slice().sort().join(",");
}

function getResult(pieces: SuperPiece[]): Result {
  console.log("getResult", pieces);
  const details = checkFromSuperPiece(pieces);
  const ps = assign2list(pieces.length, assign(details));
  const fullSet = new Set<QuantumPiece>();
  details.forEach(({ piece }) => piece.forEach((p) => fullSet.add(p)));
  const fulls = Array.from(fullSet);

  // console.log(details, fulls, ps)
  return {
    fulls,
    kinds: ps.map((p, i) => {
      const sp = pieces[i];
      if (p.length == 0) {
        return sp.filter((sp1) => fulls.indexOf(sp1) < 0);
      }
      return sp.filter((sp1) => p.indexOf(sp1) >= 0);
    }),
  };
}

const move2superPiece = ({ vec, promote }: QuantumMove) => {
  return movingBoards[promote ? 1 : 0][vec.x + "," + vec.y];
};

const filterByType: (
  moveType: MoveType,
  superPiece: SuperPiece
) => SuperPiece = (moveType, superPiece) => {
  const filter = filteringPiece(moveType);
  return superPiece.filter((kind) => filter.indexOf(kind) < 0);
};

const filteringPiece: (moveType: MoveType) => SuperPiece = (moveType) => {
  switch (moveType) {
    case MoveType.Promote:
      return ["Ki", "Ou"];
    case MoveType.NoPromote1:
      return ["Fu", "Ky", "Ke"];
    case MoveType.NoPromote2:
      return ["Ke"];
    case MoveType.Captured:
      return ["Ou"];
    default:
      return [];
  }
};
