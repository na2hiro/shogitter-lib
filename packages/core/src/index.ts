import Shogi from "./Shogi.js";
import { Species } from "./Ban.js";
import { PlayerInfo } from "./Teban.js";
import { shogitterDB } from "./ShogitterDB.js";
import XY from "./XY.js";
import { KomaInfo } from "./db/koma.js";
import * as quantumUtils from "./utils/quantumUtils.js";

export * from "./Shogi.js";
export * from "./Command.js";
export * from "./Direction.js";
export * from "./Koma.js";

export { Shogi, Species, KomaInfo, PlayerInfo, shogitterDB, XY, quantumUtils };
