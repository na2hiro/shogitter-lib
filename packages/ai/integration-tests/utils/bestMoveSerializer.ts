import { SnapshotSerializer } from "vitest";
import { BestMove } from "../../src/search/minimax";
import { Move } from "@shogitter/core/src";

export default {
  serialize(val: BestMove<Move>, config, indentation, depth, refs, printer) {
    // `printer` is a function that serializes a value using existing plugins.
    return `${printer(printBestMove(val), config, indentation, depth, refs)}`;
  },
  test(val) {
    return (
      val &&
      Object.prototype.hasOwnProperty.call(val, "moves") &&
      Object.prototype.hasOwnProperty.call(val, "score")
    );
  },
} satisfies SnapshotSerializer;

function printMove(move: Move) {
  if (move.type === "move") {
    return `${move.from[0]}${move.from[1]}${move.to[0]}${move.to[1]}${
      move.nari ? "+" : ""
    }`;
  } else if (move.type == "put") {
    return `${move.to[0]}${move.to[1]}${move.put}`;
  } else {
    return move.type;
  }
}

function printBestMove(val: BestMove<Move>) {
  return (
    val.moves.map(printMove).join(" ") +
    ` ${val.score}` +
    (val.trace?.length > 0
      ? "\n" + val.trace?.map(printBestMove).join("\n")
      : "")
  );
}
