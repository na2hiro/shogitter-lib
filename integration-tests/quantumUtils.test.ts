import {MoveType, runQuantum} from "../src/utils/quantumUtils";
import {RelXY} from "../src/XY";

describe("Quantum utils", () => {
    it("doesn't determine without move", () => {
        expect(runQuantum(0, null, MoveType.Normal, [])).toMatchSnapshot()
    })
    it("determines to KE after doing KE jump", () => {
        expect(runQuantum(0, {vec: new RelXY(-1, -2), promote: false}, MoveType.Normal, [])).toMatchSnapshot()
    })
    it("determines to 2 KEs and exhausts it", () => {
        var keMove = {vec: new RelXY(-1, -2), promote: false};
        let {kinds, fulls} = runQuantum(0, keMove, MoveType.Normal, []);
        const result = runQuantum(1, keMove, MoveType.Normal, kinds);
        expect(result).toMatchSnapshot()
        let others = runQuantum(2, null, MoveType.Normal, result.kinds);
        expect(others).toMatchSnapshot()
    })
    it("determines to HI after doing HI move and exhausts it", () => {
        expect(runQuantum(0, {vec: new RelXY(0, 2), promote: false}, MoveType.Normal, [])).toMatchSnapshot()
    })
    it("throws error to use 2 HIs", () => {
        var hiMove = {vec: new RelXY(0, 2), promote: false};
        const {kinds, fulls} = runQuantum(0, hiMove, MoveType.Normal, []);
        expect(() => {
            runQuantum(1, hiMove, MoveType.Normal, kinds);
        }).toThrow()
    })
    it("determines to KI,HI,OU after going bKekward", () => {
        expect(runQuantum(0, {vec: new RelXY(0, 1), promote: false}, MoveType.Normal, [])).toMatchSnapshot()
    })
    it("determines OU after other pieces (known piece was captured, known piece to be OU)", () => {
        const {kinds} = runQuantum(0, null, MoveType.Captured, [
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi", "Ou"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi", "Ou"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi"],
        ])
        expect(kinds[1]).toEqual(["Ou"])
    })
    it("determines OU after other pieces (known piece was captured, new piece to be OU)", () => {
        const {fulls} = runQuantum(0, null, MoveType.Captured, [
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi", "Ou"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi"],
            ["Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi"],
        ])
        // New piece will be nothing but "Ou"
        expect(fulls).toHaveLength(7)
        expect(fulls).not.toContain("Ou");
    })
    it("determines OU after other pieces (new piece was captured)", () => {
        expect(runQuantum(19, null, MoveType.Captured, [
            [ "Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi" ],
            [ "Ki", "Hi", "Ou" ],
            [ "Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi" ],
            [ "Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi" ],
            [ "Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi" ],
            [ "Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi" ],
            [ "Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi" ],
            [ "Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi" ],
            [ "Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi" ],
            [ "Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi" ],
            [ "Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi" ],
            [ "Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi" ],
            [ "Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi" ],
            [ "Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi" ],
            [ "Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi" ],
            [ "Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi" ],
            [ "Ki", "Hi", "Ou" ],
            [ "Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi" ],
            [ "Fu", "Ky", "Ke", "Gi", "Ki", "Ka", "Hi" ]
        ])).toMatchSnapshot()
    })
})
