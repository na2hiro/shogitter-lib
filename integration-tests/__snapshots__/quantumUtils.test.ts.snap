// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quantum utils determines OU after other pieces (new piece was captured) 1`] = `
Object {
  "fulls": Array [
    "Fu",
    "Gi",
    "Hi",
    "Ka",
    "Ke",
    "Ki",
    "Ky",
    "Ou",
  ],
  "kinds": Array [
    Array [
      "Fu",
      "Gi",
      "Hi",
      "Ka",
      "Ke",
      "Ki",
      "Ky",
    ],
    Array [
      "Hi",
      "Ki",
      "Ou",
    ],
    Array [
      "Fu",
      "Gi",
      "Hi",
      "Ka",
      "Ke",
      "Ki",
      "Ky",
    ],
    Array [
      "Fu",
      "Gi",
      "Hi",
      "Ka",
      "Ke",
      "Ki",
      "Ky",
    ],
    Array [
      "Fu",
      "Gi",
      "Hi",
      "Ka",
      "Ke",
      "Ki",
      "Ky",
    ],
    Array [
      "Fu",
      "Gi",
      "Hi",
      "Ka",
      "Ke",
      "Ki",
      "Ky",
    ],
    Array [
      "Fu",
      "Gi",
      "Hi",
      "Ka",
      "Ke",
      "Ki",
      "Ky",
    ],
    Array [
      "Fu",
      "Gi",
      "Hi",
      "Ka",
      "Ke",
      "Ki",
      "Ky",
    ],
    Array [
      "Fu",
      "Gi",
      "Hi",
      "Ka",
      "Ke",
      "Ki",
      "Ky",
    ],
    Array [
      "Fu",
      "Gi",
      "Hi",
      "Ka",
      "Ke",
      "Ki",
      "Ky",
    ],
    Array [
      "Fu",
      "Gi",
      "Hi",
      "Ka",
      "Ke",
      "Ki",
      "Ky",
    ],
    Array [
      "Fu",
      "Gi",
      "Hi",
      "Ka",
      "Ke",
      "Ki",
      "Ky",
    ],
    Array [
      "Fu",
      "Gi",
      "Hi",
      "Ka",
      "Ke",
      "Ki",
      "Ky",
    ],
    Array [
      "Fu",
      "Gi",
      "Hi",
      "Ka",
      "Ke",
      "Ki",
      "Ky",
    ],
    Array [
      "Fu",
      "Gi",
      "Hi",
      "Ka",
      "Ke",
      "Ki",
      "Ky",
    ],
    Array [
      "Fu",
      "Gi",
      "Hi",
      "Ka",
      "Ke",
      "Ki",
      "Ky",
    ],
    Array [
      "Hi",
      "Ki",
      "Ou",
    ],
    Array [
      "Fu",
      "Gi",
      "Hi",
      "Ka",
      "Ke",
      "Ki",
      "Ky",
    ],
    Array [
      "Fu",
      "Gi",
      "Hi",
      "Ka",
      "Ke",
      "Ki",
      "Ky",
    ],
    Array [
      "Fu",
      "Gi",
      "Hi",
      "Ka",
      "Ke",
      "Ki",
      "Ky",
    ],
  ],
}
`;

exports[`Quantum utils determines to 2 KEs and exhausts it 1`] = `
Object {
  "fulls": Array [
    "Ke",
  ],
  "kinds": Array [
    Array [
      "Ke",
    ],
    Array [
      "Ke",
    ],
  ],
}
`;

exports[`Quantum utils determines to 2 KEs and exhausts it 2`] = `
Object {
  "fulls": Array [
    "Ke",
  ],
  "kinds": Array [
    Array [
      "Ke",
    ],
    Array [
      "Ke",
    ],
    Array [
      "Fu",
      "Gi",
      "Hi",
      "Ka",
      "Ki",
      "Ky",
      "Ou",
    ],
  ],
}
`;

exports[`Quantum utils determines to HI after doing HI move and exhausts it 1`] = `
Object {
  "fulls": Array [
    "Hi",
  ],
  "kinds": Array [
    Array [
      "Hi",
    ],
  ],
}
`;

exports[`Quantum utils determines to KE after doing KE jump 1`] = `
Object {
  "fulls": Array [],
  "kinds": Array [
    Array [
      "Ke",
    ],
  ],
}
`;

exports[`Quantum utils determines to KI,HI,OU after going bKekward 1`] = `
Object {
  "fulls": Array [],
  "kinds": Array [
    Array [
      "Hi",
      "Ki",
      "Ou",
    ],
  ],
}
`;

exports[`Quantum utils doesn't determine without move 1`] = `
Object {
  "fulls": Array [],
  "kinds": Array [
    Array [
      "Fu",
      "Gi",
      "Hi",
      "Ka",
      "Ke",
      "Ki",
      "Ky",
      "Ou",
    ],
  ],
}
`;
